### PR TITLE
Automatically convert LINKXXX to LINK + use final switch

### PR DESCRIPTION
**Pull request by wilzbach**
```bash
replacements=(
"LINKdefault default_"
"LINKd d"
"LINKc c"
"LINKcpp cpp"
"LINKwindows windows"
"LINKpascal pascal"
"LINKobjc objc"
"LINKsystem system"
)
shopt -s globstar
for r in "${replacements[@]}" ; do
    w=($r)
    sed "s/${w[0]}/LINK.${w[1]}/g" -i **/*.d
done
```

### CODE REVIEW

File: src/dmd/dmangle.d [hunk 1]
```diff
@@ -543,25 +543,23 @@ public:
...
+                case LINK.windows:
+                case LINK.pascal:
+                case LINK.objc:
                     return d.ident.toChars();
-                case LINKcpp:
+                case LINK.cpp:
                     return Target.toCppMangle(d);
-                case LINKdefault:
+                case LINK.def:
+                case LINK.system:
```
Comment (wilzbach) [inline src/dmd/dmangle.d]: I wasn't sure what the behavior of `LINK.system` should be here. FWIW `LINK.system` feels like a hack.

File: src/dmd/json.d [hunk 2]
```diff
@@ -293,30 +293,35 @@ public:
...
             property(name, "windows");
             break;
-        case LINKpascal:
+        case LINK.pascal:
             property(name, "pascal");
             break;
-        default:
-            assert(false);
+        case LINK.objc:
+            property(name, "objc");
```
Comment (wilzbach) [inline src/dmd/json.d]: `final switch` for the win!

File: src/dmd/target.d [hunk 3]
```diff
@@ -478,7 +484,7 @@ struct Target
      */
     extern (C++) static LINK systemLinkage()
     {
-        return global.params.isWindows ? LINKwindows : LINKc;
+        return global.params.isWindows ? LINK.windows : LINK.c;
```
Comment (wilzbach) [inline src/dmd/target.d]: This is why `LINK.system` feels like a hack. Should `systemLinkage` be called everywhere instead of using `LINK.system`?

Comment (wilzbach) [reply src/dmd/target.d]: Hmm `LINK.system` is only set once by the parser:

https://github.com/wilzbach/dmd/blob/089241b1359d1a20f19dd1b589ff0922854b16bc/src/dmd/parse.d#L2208-L2211

**Review by JinShil — CHANGES_REQUESTED** [review]

File: src/dmd/astbase.d [hunk 4]
```diff
@@ -6278,24 +6278,23 @@ struct ASTBase
 
     extern (C++) static const(char)* linkageToChars(LINK linkage)
     {
-        switch (linkage)
+        final switch (linkage)
         {
-        case LINKdefault:
+        case LINK.def:
```
[the remark this answers is not in the document]
Comment (wilzbach) [reply src/dmd/astbase.d]: I didn't change that:

https://github.com/dlang/dmd/blob/d9421bde847c24a8f8402b885e594859c394513c/src/dmd/globals.d#L415-L425

[the remark this answers is not in the document]
Comment (JinShil) [reply src/dmd/astbase.d]: Yes, I know.  But this is the opportune time to change it.  `def` is a pretty lousy name, and `default_` is also used elsewhere.  While we're refactoring these we have the opportunity to make them all consistent.

[the remark this answers is not in the document]
Comment (JinShil) [reply src/dmd/astbase.d]: I'll submit a change for #7737 if you do this one.

[the remark this answers is not in the document]
Comment (wilzbach) [reply src/dmd/astbase.d]: Ok changed, I can change #7737 too - I don't mind.

> But this is the opportune time to change it.

Shouldn't we rename the enum to PascalCase too then?

[the remark this answers is not in the document]
Comment (wilzbach) [reply src/dmd/astbase.d]: > the same should apply for #7737

https://github.com/dlang/dmd/pull/7745

File: src/dmd/backend/ty.d [hunk 5]
```diff
@@ -74,7 +74,7 @@ enum
     TYifunc             = 0x2E, // interrupt func
     TYptr               = 0x33, // generic pointer type
     TYmfunc             = 0x37, // NT C++ member func
-    TYjfunc             = 0x38, // LINKd D function
+    TYjfunc             = 0x38, // LINK.d D function
```
Comment (ibuclaw) [inline src/dmd/backend/ty.d]: I wonder if the `j` stands for Jupiter also.

File: src/dmd/json.d (hunk 2 above)
Comment (ibuclaw) [reply src/dmd/json.d]: Add test. ;-)

**Review by ibuclaw — APPROVED** [review]
Ok

**Review by jacob-carlborg — APPROVED** [review]
Approved assuming the test that Iain mentioned is added.

**Comment by wilzbach** [comment]
> final switch for the win!
> Add test. ;-)

Urgh. The JSON output didn't print `linkage` information at all. Glad that we have Coverage annotation :)
-> it was trivial to fix & I added a test for it.

**Review by JinShil — APPROVED** [review]

**Pull request merged**